### ISSUE TITLE
RFC: Where to go with "ontick" and friends?

### ISSUE THREAD

**Issue by hmans** [hmans/three-elements#35]
three-elements' `<three-game>` root element implements a gameloop that you can hook into using four different ticker callbacks:

- `tick` (executed every tick, no exceptions; ie. 60 times per second)
- `latetick` (exactly like `tick`, but guaranteed to run after all `tick` listeners are done executing; a common game loop pattern)
- `frametick` (executed only on ticks that are going to render a frame)
- `rendertick` (executed _while_ rendering a frame; you can use it to hook in your custom rendering code)

Right now, in HEAD -- which is going to turn into a 0.3.0 release soon -- there are three ways to hook into these events:

- directly attach callbacks to the `three-game` element's `emitter`, eg. `game.emitter.on("tick", () => ...)`
- set the `ontick`, `onlatetick`, `onframetick` or `onrendertick` _property_ of any three element to a function, eg. `mesh.ontick = () => ...`
- set the `ontick`, `onlatetick`, `onframetick` or `onrendertick` _attribute_ of any three element to the string representation of a function body, eg. `<three-mesh ontick="this.object.rotateZ(0.01)">`

An earlier version of three-elements was using DOM events (CustomEvent instances sent though dispatchEvent), but this unfortunately very quickly ruined performance once more than just a tiny handful of elements were using this mechanism.

The current setup works well and performs nicely, _but_ the API has some issues that I'd like to resolve:

- The fact that you can set `ontick` on an element also implies that the element dispatches a "tick" event (see [this Twitter thread](https://twitter.com/hmans/status/1354180092814098432)). This is not only an expected pattern, but is also an assumption made by some frameworks (eg. [HyperApp](https://github.com/jorgebucaran/hyperapp)) that implement their own handling of `on*` attributes.
- We like to avoid surprises, but it's certainly surprising that there is a `ThreeGame.emitter`, but no `ThreeElement.emitter`. A lot of people on Twitter have voiced the opinion that it's important to be able to attach multiple event listeners to the same event, which would currently require you to attach to the `three-game` element's emitter directly. (Note: I don't believe this is an important use case that we need to support; ticker callbacks are not UI events, after all.)

If anyone has any suggestions on how to move forward with this API, let's hear them!

The options I'm seeing right now are the following:

- **Leave things as they are now**, at least for the time being, and live with the fact that there's going to be some surprising behavior in some frameworks (ie. in lit-html, you can do `.ontick=${fn}`, but you can't do `@tick=${fn}`. It's not great, but also not the end of the world.)
- **Rename these properties/attributes** to something that doesn't imply that they're reflecting DOM events, for example `pertick`, or `tickfn`, or `tickcallback`, or...?
- **Go back to using real DOM events**, which is only acceptable if we find a way to not completely thrash performance with them.
- **Remove element-specific ticker event listeners**, have only `three-game` emit ticks as proper DOM Events, and require that all per-frame code hook into that and only that. A pretty radical idea that would make things completely non-surprising at the cost of API approachability.

**Comment by daKmoR** [comment]
> Go back to using real DOM events, which is only acceptable if we find a way to not completely thrash performance with them.

how did you dispatch events? if you dispatch without bubble & compose then I would be surprised if it trashes performance? 🤔  - any way to test this?

**Comment by hmans** [comment]
> how did you dispatch events? if you dispatch without bubble & compose then I would be surprised if it trashes performance? 🤔 - any way to test this?

Bubbling was disabled (of course.)

Consider that this happens 60 times per second (or more, if you're on a higher-frequency monitor). If you have 1 element that hooks into this, that's 60 events per second -- not a big problem. But if you have 10, you're already at 600. If you have 100, at 6000.

With a normal EventEmitter pattern, 100 (or event many more) of these are not a problem at all, but going through `dispatchEvent` quickly becomes problematic (starting with the hight number of objects created and ending with the relative overhead of the event dispatching mechanism.)

**Comment by pookage** [comment]
If they're named `ontick` then I would absolutely expect a `tick` event to be fired that I could listen for traditionally - so I'd advocate either for renaming or finding a way to make it work.

> Remove element-specific ticker event listeners, have only three-game emit ticks as proper DOM Events, and require that all per-frame code hook into that and only that. A pretty radical idea that would make things completely non-surprising at the cost of API approachability.

This feels like a good approach?

**Comment by hmans** [comment]
> This feels like a good approach?

I need to think about it some more. Semantically, it feels like a good change, since it is not the scene objects that are doing the ticking, they're just _reacting_ on it. But the impact this would have on the overall approachability of the library worries me.

In the current setup, the most simple thing you can do to make a mesh spin is to just do this on the mesh element:

```html
<three-mesh ontick="this.object.rotateZ(0.01)">
```

When you add a framework on top -- let's use lit-html as an example -- you can do this, which sidesteps all the string attribute stuff:

```html
<three-mesh .ontick=${(dt, element) => element.object.rotateZ(0.01)}>
```

And of course you can do it fully in JS-land, once you have a reference to the element:

```js
element.ontick = () => element.object.rotateZ(0.01)
```

If we implement a change like the one we're talking about, the user would be _forced_ into JS-land to do something like the following, and do their own lifecycle management:

```js
element.game.addEventListener("tick", (e) => e.target.object.rotateZ(0.01))
``` 

tl;dr I'm not a fan, but I will think on it some more.

**Comment by daKmoR** [comment]
I stand corrected... only dispatching the event makes the code already sooooo much slower (~99% slower 😱 )

https://jsbench.me/qmkkfcfpcw/2

Knowing that my gut feeling would vote for an callback name + using a lib to set it as a property

```html
<three-mesh .tickCallback=${(dt, element) => element.object.rotateZ(0.01)}>
```

and if you really want it to work with attributes alone... then it might be a good idea to follow the `data-*` or svg like `stroke-width` pattern to make it clearly different then the property. I would even go as far as adding a separate attribute like `use-unsafe-attributes` (name to be discussed) to enable this behavior

```html
<three-mesh use-unsafe-attributes tick-callback="this.object.rotateZ(0.01)">
```

**Comment by hmans** [comment]
> if you really want it to work with attributes alone

There might be a misunderstanding here. This is certainly not the case. I _also_ want to support setting ticker callbacks as attributes; I certainly don't want to make it the _only_ option.

> vote for an callback name + using a lib to set it as a property

This is already very close to (if not the same as?) what I'm doing today, except of course for using `on*` names, which would need to be changed (see the second option from my list at the bottom of the original post.) If I had to make a decision right now, this would be my preferred option.

**Comment by hmans** [comment]
Current favorite, after some thinking: `update`, `update.late`, `update.frame`, `update.render` and later `update.[customname]` (when we allow users to expose additional ticking stages.)

The dot-less version being the default per-rAF tick.

Doesn't have to be `update`, can also be `tick`, or `callback`, or something else, as long as it doesn't start with `on`.

**Comment by hmans** [comment]
I'm getting ready to implement this, but I've moved away from the "dotty" syntax. The goal is to implement the following properties on ThreeElement to replace `ontick`, `onlatetick` etc.:

```
tickUpdate
tickLateUpdate
tickFrame
tickRender
```

And also the following attributes that map to the above properties:

```
tick-update
tick-late-update
tick-frame
tick-render
```

This checks the following boxes:

- [x] The naming is aligned with custom element conventions (kebab-case attributes mapping to camelCase properties)
- [x] It no longer causes confusing by properties/attributes' `on*` naming suggesting that there are DOM events of the same names that you could listen to (there are not, and they're not coming back as long as dispatching a massive amount of DOM events is multiple orders of magnitudes slower than using an EventEmitter pattern, sorry)
- [x] It doesn't actually change any of the functionality that is on place now; we're just renaming things
- [x] It leaves the door open for future tick callback types that user may be adding, like `tick-physics` or `tick-animation`, if we ever do such a thing on a property/attribute level
- [x] It doesn't change the fact that you can (and always will be able to) hook into ThreeGame's emitter directly, which you will most likely do _instead of_ setting element properties or attributes in non-trivial, JS-driven projects.

**Comment by hmans** [comment]
The above patch implements this, with the addition of `tick` being available as a shortcut for `tickUpdate`, since this is the most common callback to hook into. Closing this, thanks to everyone who gave feedback!

**Comment by hmans** [comment]
I've done another pass on this, the properties are now `tick`, `lateTick`, `frameTick` and `renderTick`, and the attributes `tick`, `late-tick`, `frame-tick` and `render-tick`. I think it's good now!
